docs/proposals/block-storage-time-series-deletion.md

```diff
@@ -0,0 +1,255 @@
+
+### Permanently deleting the data
+
+The proposed approach is to perform the deletions from the compactor. A new background service inside the compactor called _DeletedSeriesCleaner_ can be created and is responsible for executing the deletion.
+
+#### Processing
+
+
+This will happen after a grace period has passed once the API request has been made. By default this should be 24 hours. A background task can be created to process the permanent deletion of time series. This background task can be executed each hour.
+
+To delete the data from the blocks, the same logic as the [Bucket Rewrite Tool](https://thanos.io/tip/components/tools.md/#bucket-rewrite
+) from Thanos can be leveraged. This tool does the following: `tools bucket rewrite rewrites chosen blocks in the bucket, while deleting or modifying series`. The tool itself is a CLI tool that we won’t be using, but instead we can utilize the logic inside it. For more information about the way this tool runs, please see the code [here](https://github.com/thanos-io/thanos/blob/d8b21e708bee6d19f46ca32b158b0509ca9b7fed/cmd/thanos/tools_bucket.go#L809).
+
+The compactor’s _DeletedSeriesCleaner_ will apply this logic on individual blocks and each time it is run, it creates a new block without the data that matched the deletion request. The original individual blocks containing the data that was requested to be deleted, need to be marked for deletion by the compactor.
+
+While deleting the data permanently from the block storage, the `meta.json` files will be used to keep track of the deletion progress. Inside each `meta.json` file, we will add a new field called `tombstonesFiltered`. This will store an array of deletion request id's that were used to create this block. Once the rewrite logic is applied to a block, the new block's `meta.json` file will append the deletion request id(s) used for the rewrite operation inside this field. This will let the _DeletedSeriesCleaner_ know that this block has already processed the particular deletions requests listed in this field. Assuming that the deletion requests are quite rare, the size of the meta.json files should remain small.
+
+The _DeletedSeriesCleaner_ can iterate through all the blocks that the deletion request could apply to. For each of these blocks, if the deletion request ID isn't inside the meta.json `tombstonesFiltered` field, then the compactor can apply the rewrite logic to this block. If there are multiple tombstones that are currently being processing for deletions and apply to a particular block, then the _DeletedSeriesCleaner_ will process both at the same time to prevent additional blocks from being created. If after iterating through all the blocks, it doesn’t find any such blocks requiring deletion, then the `Pending` state is complete and the request progresses to the `Processed` state.
```

Comment:
If the tombstone file has a `maxTime` which is in the future, when do you consider all blocks are processed? Do we need to wait until all blocks are processed and the `maxTime` is passed?

Reply:
After taking a break, I'm getting back to working on the implementation of this proposal. 

I was following the chunks series deletion API implementation, which I don't think supports deletes with end times in the future. However, the way you mention it here could work as well.  Need to check that the end time has passed before moving the deletion request to the processed state. 

If we allow deletion requests with the end times in the future, one downside is that if the deletion period end time is really far into the future, then the tombstone query time filtering will result in slower performance for a long time until the tombstone end time has passed.  For example, let's say we would like to delete a certain metric 1 month into the future, then all queries will have to be filtered for that metric the entire month.

Reply:
It is not clear that how to process multiple tombstones on the same block. As two tombstones may enter processed state at different time, so one block might be rewritten first by one tombstone then another tombstone can only apply to the new block.

Reply:
If there are multiple tombstones that are past the pending state, then we can rewrite a single block once using multiple tombstones at the same time. If there is some tombstones in a pending state and some in processing, then it is possible the block will get overridden once by one tombstone and then the new block gets overridden again once the pending tombstone begins processing. 

My plan was that the compactor will process the permanent deletions at a configurable interval, `SeriesDeletionCleanupInterval`. So the deletions will not necessarily begin as soon as the pending period passes, there could be some waiting, which depends on how much time you will set it to. So if lets say you set `SeriesDeletionCleanupInterval` = 12 hours, and in the last 12 hours X tombstones went from pending to processing state, it will only rewrite each block once using the X tombstones at the same time (assuming the X tombstones would apply to the block).  So making multiple delete requests close to each other will require less processing than doing them far apart.